### DISCUSSION TITLE
아이템 20. 추상 클래스보다는 인터페이스를 우선하라

### DISCUSSION THREAD

**Discussion by coalong** [Study-2-Effective-Java/Effective-Java#40] [3. 과제] [answered]
# 아이템 20. 추상 클래스보다는 인터페이스를 우선하라

## TL;DR   
- 일반적으로 다중 구현을 할 때에는 인터페이스가 가장 적합하다!   

- 복잡한 인터페이스라면 **골격 구현 방법**을 꼭 고려해보자!   

- 가능하다면, 골격 구현은 인터페이스의 default 메서드 제공해야 하며, 그 인터페이스를 모든 곳에서 사용하는 것이 좋다.   
(인터페이스에 걸려 있는 구현상의 제약(e.g. default method의 Object 메서드 재정의 불가) 때문에 골격 구현을 추상 클래스로 제공하는 경우가 더 흔하다.)

## 자바의 다중 구현 메커니즘

<details>
<summary>자바가 제공하는 다중 구현 메커니즘은 인터페이스와  추상 클래스, 이렇게 두가지다.</summary>
<div markdown="1">     
   
- Interface   

    - 공통 기능을 명세하는 목적   
    - Abstract Method 를 정의해놓고 구현(implements)하는 Class에서 각 기능들을 Overriding하여 여러가지 형태로 구현할 수 있기에 다형성과 연관   
    - HAS - A    “~을 할 수 있는”   
    
- Abstract Class
    - 부모와 자식 즉, 계층이 있는 상속 관계
    - 자식 Class 들 간에 공통 기능을 각각 구현하고, 확장시킨다.
    - IS - A   “~이다.”
</div>
</details>

<details>
<summary>자바 8부터 인터페이스도 디폴트 메서드(default method)를 제공할 수 있어 이제는 두 메커니즘 모두 인스턴스 메서드를 구현 형태로 제공할 수 있다.</summary>
<div markdown="1">    
   
- **인터페이스의 디폴트 메서드 ❓**
    - Java 7 이전에는 추상 메서드, 상수 정의만 가능했다.
    - Java 8 부터는 default method 와 static method 도 정의 할 수 있다.
    - Default Method 란?
        - 인터페이스에서 **기본 구현**을 갖는 메서드이다.
        - 인터페이스의 메서드 중 공통 기능 명세 
        - 인터페이스의 유연성과 하위호환성 등 유지보수를 할 수 있도록 (참고자료)
            [참고자료 : 인터페이스의 디폴트 메서드](https://jeong-pro.tistory.com/209) 
</div>
</details>        

### 타입 제약 문제!!
- 자바는 단일 상속만 지원하니, 추상 클래스로는 확장에 용이하지 않다.
    - **추상 클래스**가 정의한 타입을 구현하는 클래스는 **반드시 추상 클래스의 하위 클래스**가 되어야 한다.
- **인터페이스**가 선언한 메서드를 모두 정의하고 그 일반 규약을 잘 지킨 클래스라면 **다른 어떤 클래스를 상속했든 같은 타입**으로 취급된다.

## 인터페이스로 구현하면 이점

<details>
<summary>기존 클래스에도 손쉽게 새로운 인터페이스를 구현해넣을 수 있다.</summary>
<div markdown="1">    

- 인터페이스가 요구하는 메서드를 추가하고, 클래스 선언에 implements 구문만 추가한다.  

- 예시로, 자바 플랫폼에서도 Comparable, Iterable, AutoCloseable 인터페이스가 새로 추가됐을 때 표준 라이브러리의 수많은 기존 클래스가 이 인터페이스들을 구현한 채 릴리스됐다.  

- 반면에, 기존 클래스 위에 새로운 추상 클래스를 끼워넣기는 어려운게 일반적이다.  

- 두 클래스가 같은 추상 클래스를 확장하길 원한다면, 그 추상 클래스는 계층구조상 두 클래스의 공통 조상이어야 한다. 이러한 방식은 클래스 계층구조에 커다란 혼란을 일으킨다.    

- 새로 추가된 `추상 클래스의 모든 자손이 적절하지 않은 상황에서도 강제로 이를 상속` 하게 되는 것이다.

</div>
</details>    

<details>
<summary>인터페이스는 믹스인(mixin) 정의에 안성맞춤이다.</summary>
<div markdown="1">    

- 믹스인이란?     
    - `주된 기능에 선택적 기능을 혼합(mixed in)` 하는 것!   

- 예를 들어 `Comparable` 인터페이스는 자신을 구현한 클래스의 인스턴스들끼리는 `순서를 정할 수 있다고 선언하는 믹스인` 인터페이스이다.   

- 반면에, 추상 클래스로는 믹스인을 정의할 수 없다. `클래스는 다중 상속이 안되고`, 클래스 계층구조에는 믹스인을 삽입하기에 합리적인 위치가 없다.

</div>
</details>    

<details>
<summary>인터페이스로는 계층구조가 없는 타입 프레임워크를 만들 수 있다.</summary>
<div markdown="1">       

- 인터페이스 예시
    
```java
    // 가수 인터페이스
    public interface Singer {
    	AudioClip sing(Song s);
    }
    
    // 작곡 인터페이스
    public interface Songwriter {
    	Song compose(int chartPosition);
    }
    
    // 계층구조 없이 설계할 수 있다.
    // Singer 인터페이스와 Songwriter 인터페이스를 모두 구현한 SingerSongwriter 인터페이스 정의
    // 위에 두 인터페이스 메서드 뿐만 아니라 새로운 메서드까지 추가한 제 3의 인터페이스 정의 가능
    public interface SingerSongwriter extends Singer, Songwriter {
    	AudioClip strum();
    	void actSeneitive();
    }
```  

    
- 위의 코드와 같은 구조를 `클래스로 만들려면 가능한 조합 전부`를 각각의 클래스로 정의한 **고도비만 계층**구조가 만들어질 것이다.   

- 속성이 n개라면 지원해야 할 조합의 수는 2^n개나 된다. 흔히 조합 폭발(combinatorial explosion)이라 부르는 현상이다.    

- 거대한 클래스 계층구조에는 공통 기능을 정의해놓은 타입이 없으니, 자칫 매개변수 타입만 다른 메서드들을 수없이 많이 가진 거대한 클래스를 낳을 수 있다.
    
```java
    public abstract class Singer {
    	abstract AudioClip sing(Song s);
    }
    public abstract class SongWriter {
        abstract Song compose(int chartPosition);
    }
    public abstract class SingerSongWriter {
        abstract AudioClip strum();
        abstract void actSensitive();
        abstract Song Compose(int chartPosition);
        abstract AudioClip sing(String s);
    }
```
</div>
</details>  
    
- 래퍼 클래스 관용구(아이템 18)와 함께 사용하면 인터페이스는 기능을 향상시키는 안전하고 강력한 수단이 된다.
    - 타입을 추상 클래스로 정의해두면 그 타입에 기능을 추가하는 방법은 상속뿐이다.
    - 상속해서 만든 클래스는 래퍼 클래스보다 활용도가 떨어지고 깨지기는 더 쉽다.

## 인터페이스의 디폴트 메서드(default method)

- 인터페이스의 메서드 중 `구현 방법이 명백한 것`이 있다면, 디폴트 메서드를 제공하면 구현하는데 편리함을 준다.
- 기존 인터페이스를 구현하는 클래스에서 `자동으로 인터페이스에 추가된 새로운 디폴트 메서드를 상속`받아서 `기존의 코드를 수정하지 않고 기능을 확장`할 수 있다.
    - e.g. [Iterable 인터페이스](https://coder-in-war.tistory.com/m/entry/Java-27-Iterable%EA%B3%BC-Iterator-%EC%9D%B8%ED%84%B0%ED%8E%98%EC%9D%B4%EC%8A%A4)의 디폴트 메서드  forEach() 
- 디폴트 메서드를 제공할 때는 **@implSpec 자바독 태그**를 붙여 상속하려는 사람을 위해 **문서화**해야 한다. (아이템 19)

###  디폴트 메서드의 제약   

1. Object 의 equals와 hashcode, toString 같은 메서드를 `디폴트 메서드로 제공해서는 안된다.`
        
    - [Java8: java.lang.Object의 메서드에 대한 디폴트 메서드를 정의하는 것이 금지된 이유](https://stackoverflow.com/questions/24016962/java8-why-is-it-forbidden-to-define-a-default-method-for-a-method-from-java-lan?noredirect=1&lq=1)
        
2. 인터페이스는 `인스턴스 필드`와 `public이 아닌 정적 멤버`를 가질 수 없다. (인터페이스의 기본 접근 제어자는 public)
        (단, private 정적 메서드는 예외다.)
3. 우리가 만들지 않은 인터페이스에는 디폴트 메서드를 추가할 수 없다.

## 인터페이스와 추상 골격 구현(skeletal implementation) 클래스

- `인터페이스`와 `추상 골격 구현(skeletal implementation)` 클래스를 함께 제공하여 `인터페이스와 추상 클래스의 장점을 모두` 취할 수 있다.  
    → 추상 클래스처럼 구현을 도와주는 동시에, 추상클래스로 타입을 정의할 때 따라오는 심각한 제약에서 자유롭다.
    → 이러면 단순히 골격 구현을 확장하는 것만으로 이 인터페이스를 구현하는 데 필요한 대부분의 일이 완료된다.
- 이것이 `템플릿 메서드 패턴`을 적용한 방법이다!
    - [템플릿 메서드 패턴이란?](https://steady-coding.tistory.com/384)
- 관례상 인터페이스 이름이 interface라면 그 골격 구현 클래스의 이름은 AbstractInterface로 짓는다.
    - e.g. AbstractCollection, AbstractSet, AbstractList, AbstractMap

### List 인터페이스와 AbstractCollection 추상 클래스를 확장하여 AbstractInterface 골격 구현 클래스 생성
```java
    public abstract class AbstractList<E> extends AbstractCollection<E> implements List<E> {
        ...
    }
```
    
- 제대로 설계했다면 골격 구현은 그 인터페이스로 나름의 구현을 만들려는 프로그래머의 일을 상당히 덜어준다.  
(위의 예시로 AbstractList 골격 구현을 활용하면 보다 쉽게 List 인터페이스를 구현할 수 있다.)

<details>
<summary>AbstractList 예시</summary>
<div markdown="1">    

- `완벽히 동작하는 List 구현체를 반환`하는 정적 팩터리 메서드로, `AbstractList 골격 구현`으로 활용
    - 이 코드는 골격 구현의 힘을 잘 보여주는 인상적인 예시 이다.
    
    ```java
    public class Example {
    
        static List<Integer> intArrayAsList(int[] a) {
            // Null Check 를 하여 미리 예외를 처리 (Fail-Fast)
            Objects.requireNonNull(a);
    
            // 다이아몬드 연산자를 이렇게 사용하는 건 자바 9부터 가능하다.
            // 더 낮은 버전을 사용한다면 <Integer>로 수정하자.
            // 익명클래스 형태 사용
            return new AbstractList<>() {
    
                @Override public Integer get(int i) {
                    return a[i]; //오토박싱(아이템 6)
                }
    
                @Override public Integer set(int i, Integer val) {
                    int oldVal = a[i];
                    a[i] = val;    //오토언박싱
                    return oldVal;   //오토박싱
                }
    
                @Override public int size() {
                    return a.length;
                }
            };
        }
    
        public static void main(String[] args) {
            List<Integer> integers = Example.intArrayAsList(new int[]{1, 2, 3});
        }
    }
    ```
    > 이 예시 코드는 int 배열을 받아 Integer 인스턴스의 리스트 형태로 보여주는 어댑터(Adapter)이기도 하다.

</div>
</details>    

<details>
<summary>Map.Entry 예시</summary>
<div markdown="1">    

- 간단한 예로, Map.Entry 인터페이스를 살펴보자. getKey, getValue는 확실히 기반 메서드이며, 선택적으로 setValue도 포함할 수 있다.
- 이 인터페이스는 equals와 hashCode의 동작 방식도 정의해놨다.
- Objects 메서드들은 디폴트 메서드로 제공해서는 안 되므로, 해당 메서드들은 모두 골격 구현 클래스에 구현한다.
- toString도 기반 메서드를 사용해 구현해놨다.
        
```java
public abstract class AbstractMapEntry<K, V> implements Map.Entry<K, V> {
        
        //변경 가능한 엔트리는 이 메서드를 반드시 재정의해야 한다.
        @Override public V setValue(V value) {
        	throw new UnsupportedOperationException();
        }
        
        //Map.Entry.equals의 일반 규약을 구현한다. 
        @Override public boolean equals(Object o) {
        	if (0 == this)
        		return true;
        	if (!(o instanceof Map.Entry))
        		return false;
        	Map.Entry<?, ?> e = (Map.Entry) o;
        		return Objects.equals(e.getKey(), getKey()) && Objects.equals(e.getValue(), getValue());
        }
        
        // Map.Entry.hashCode의 일반 규약을 구현한다.
        @Override public int hashCode() {
        	return Objects.hashCode(getKey())^ Objects.hashCode(getValue());
        }
        
        @Overrride public String toString() {
        	return getKey() + "=" + getValue();
        	}
        }
```
        
✅ Map.Entry 인터페이스나 그 하위 인터페이스로는 이 골격 구현을 제공할 수 없다.
             디폴트 메서드는 equals, hashCode, toString 같은 Object 메서드를 재정의할 수 없기 때문이다.
  
</div>
</details>    
   
      
- **골격 구현 작성 방법**
    1. 인터페이스에서 기반 메서드(공통 기능)를 선정.
    2. 기반 메서드는 디폴트 메서드로 기능 명세.
    3. 디폴트 메서드로 제공하지 않은 메서드를 해당 인터페이스를 구현하는 골격 구현 클래스에서 작성
    4. 설계, 문서화 작성 필요!
    

## **시뮬레이트한 다중 상속**

- 골격 구현 클래스를 우회적으로 이용할 수 있다. 인터페이스를 구현한 클래스에서 해당 골격 구현을 확장한 private 내부 클래스를 정의하고, 각 메서드 호출을 내부 클래스의 인스턴스에 전달하는 것이다. (아래 예시 코드 참고)
-  #38 와 비슷한 이 방식을 **시뮬레이트한 다중 상속**(simulated multiple inheritance)이라 하며, 다중 상속의 많은 장점을 제공하는 동시에 단점은 피하게 해준다.

## 골격 구현과 **시뮬레이트한 다중 상속** 예시 코드
<details>
<summary>골격 구현 예시 코드 (템플릿 메서드 패턴)</summary>
<div markdown="1">    

```java
// 인터페이스
public interface Phone {
  void booting();
  void greeting();
  void shutdown();
  void process();
}

// 추상 골격 구현 클래스 (보통 Abstract[interface 이름] 네이밍을 사용)
public abstract class AbstractPhone implements Phone {

    //같은 동작을 하는 메서드(기반 메서드)를 정의
    @Override
    public void booting() {
        System.out.println("booting ...");
    }

    @Override
    public void shutdown() {
        System.out.println("shut down ...");
    }

    @Override
    public void process() {
        booting();
        greeting();
        shutdown();
    }
}

public class IPhone extends AbstractPhone {

  // 기반 메서드가 아닌 메서드 구현
  @Override
  public void greeting() {
    System.out.println("I am iPhone");
  }
}

public class GalaxyPhone extends AbstractPhone {
  @Override
  public void greeting() {
    System.out.println("I am galaxy phone");
  }
}

```
</div>
</details>    

<details>
<summary> 시뮬레이트한 다중 상속(simulated multiple inheritance)</summary>
<div markdown="1">    

- PhoneManufacturer 라는 클래스를 상속해야할 상황이 생긴다면? 다중상속이 안되는데 어떻게 할까?
    
```java
public class PhoneManufacturer {
    public void printManufacturer() {
        System.out.println("Made by Apple");
    }
}
```
    
```java
// 골격 구현을 확장한 클래스 정의
public class InnerAbstractPhone extends AbstractPhone {
    @Override
    public void greeting() {
        System.out.println("I am iPhone");
    }
}
```
    
```java
// 시뮬레이트한 다중 상속
public class IPhone extends PhoneManufacturer implements Phone{
    
    // 골격 구현을 확장한 클래스를 private 내부 클래스로 정의 (합성, Composition)
    private final InnerAbstractPhone innerAbstractPhone = new InnerAbstractPhone();
    
    // 메서드 호출을 내부 클래스의 인스턴스에 전달.
    @Override
    public void booting() {
        innerAbstractPhone.booting();
    }
    @Override
    public void greeting() {
        innerAbstractPhone.greeting();
    }
    @Override
    public void shutdown() {
        innerAbstractPhone.shutdown();
    }
    @Override
    public void process() {
        printManufacturer();
        innerAbstractPhone.process();
    }
}
```
[자세한 골격 구현 예시 코드](https://dzone.com/articles/favour-skeletal-interface-in-java) 
</div>
</details>    
    

## 단순 구현

- 단순 구현도 상속을 위해 인터페이스를 구현한 것이지만, 추상 클래스가 아니란 점이 다르다.
- AbstractMap.SimpleEntry가 좋은 예다.
- 쉽게 말해 `동작하는 가장 단순한 구현`이다. 이러한 단순 구현은 그대로 써도 되고, `필요한 맞게 확장`해도 된다.

**Comment by JoisFe** [comment, marked as answer]
추상클래스보단 인터페이스를 우선시 하는 이유에 대해 알 수 있게된 좋은 정리 감사합니다.
해당 글과 무관할 수 있으나
저는 ```자바의 인터페이스가 ADT에 가장 잘 대응 되는 개념```이라는 표현이 가장 와 닿았습니다. 

### ADT(Abstract Data Type, 추상 데이터 타입) : 세부 사항에서 벗어나 추상적으로 정의한 데이터 타입.

저는 처음 data type (자료형 이라고 부르는)을 접했을 때 단순히 자바에서는 크게 2가지 기본 데이터 타입(int, boolean, 등등), 참조형 데이터 타입이 있다 정도로 이해하고 넘어갔습니다.

하지만 객체지향의 사실과 오해라는 책에서 추상화를 설명하는 부분에서 
- 추상화 : 현실에서 출발하되 불필요한 부분을 도려내가면서 사물의 놀라운 본질을 드러나게 하는 과정 --> 불필요한 부분을 무시함으로써 현실에 존재하는 복잡성을 극복

- 데이터 타입 : 메모리 안에 저장된 데이터의 종류를 분류하는 데 사용하는 메모리 집합에 관한 메타데이터로 추상화 한 결과 --> 컴퓨터의 메모리에서는 0과 1만 존재 하여 타입이라는 질서가 존재하지 않는데 데이터의 종류에 맞게 추상화 한 결과

- 객체 : 객체지향 프로그램에서 객체를 일종의 데이터 처럼 사용하는데 객체를 타입에 따라 분류하고 그 타입에 이름을 붙이는 것은 결국 프로그램에서 사용할 데이터 타입을 선언하는 것과 같음 --> 객체 또한 추상화의 결과
### 단 객체는 데이터는 아니다  객체는 행동이 중요한 것이고 상태는 행동의 결과로 초래된 부수효과를 쉽게 표현하기 위해 도입한 추상적인 개념일 뿐 (객체와 관련된 내용은 깊게 이야기 하진 않겠습니다.)

- 클래스 : 객체지향 프로그래밍 언어에서 정적인 모델은 클래스를 이용해 구현 <-- 타입을 구현하는 가장 보편적인 방법이 클래스 

주의할 점은 클래스와 타입은 동일한 것이 아님 타입은 객체를 분류하기 위해 사용하는 개념

반면 클래스는 단지 타입을 구현할 수 있는 여러 구현 메커니즘 중 하나일 뿐

이렇듯 클래스를 타입을 구현하는 방법 중 하나이고 다시 본론으로 돌아오면 자바의 인터페이스가 ADT에 가장 잘 대응 되는 개념이라는 표현이 객체지향의 사실과 오해를 읽고나서 더 이해가 되기 시작했습니다.

세부 사항에서 벗어나 추상적으로 정의한 데이터 타입인 ADT는 매우 다양한 표현 방법이 있습니다.
대표적으로 대상이 포함하는 데이터 속성, 제한 등을 서술하는 방법이 있습니다.

대표적으로 자료구조인 List의 ADT를 표현해보면

ADT '리스트'
>  - 작업: 
>    >    - i번쨰 자리에 새 원소 넣기
>    >    - 원소 X 찾기
>    >    - i번째 자리의 원소 삭제하기


이 것을 자바의 인터페이스로 표현해 보겠습니다.

``` java
public interface ListInterface {
  void add (int i, Object x);
  Object get(int i);
  void remove(int i);
}
```

자바의 인터페이스 코드로 위 리스트 자료구조에 대한 ADT 서술을 잘 표현하고 있습니다.
- 각 작업을 제공
- 각 작업의 리턴 타입
- 각 작업의 파라미터의 수와 타입
이렇게 필요한 정보도 나타나 있습니다.

결과적으로
1. 해당 인터페이스의 구현체들이 이러한 정보들을 잘 지키면서 구현하게 함
2. 해당 리스트를 사용하는 사용자 프로그램 입장에서 내부가 어떻게 구현되어있는지 알 필요 없이 추상적 수준에서 리스트를 사용할 수 있게 함

즉 자바의 인터페이스는 추상화를 하는 이유를 결과로 내놓고 있고 ADT에 가장 잘 대응되는 개념인 이유도 알게 되었습니다.

### reference
1. 객체지향의 사실과 오해 (조영호 지음) 위키북스
2. 쉽게 배우는 자료 구조 with 자바 (문병로 지음) 한빛아카데미

@coalong 님께서 작성하신 주제와 조금 벗어난 내용을 왜 comment에 남겼을까? 라는 생각을 하실수도 있지만
저는 ```추상클래스보다 인터페이스를 우선하라```는 아이템 주제의 이유가 인터페이스는 추상클래스보다 더 추상화된 개념이고
그로 인해 추상화의 장점 (불필요한 부분을 무시함으로써 현실에 존재하는 복잡성을 극복) 으로 인해 좀 더 객체 지향적 프로그래밍을 할 수 있게 되기 때문이라고 생각합니다. 이러한 이유 중 일부가 위 정리로 생각이 들어 글을 남기게 되었으니 이렇게 생각할 수 있구나 참고만 해주시면 좋을 것 같고 오개념이 존재하거나 의문이 드는 부분 있다면 comment로 남겨주시면 감사하겠습니다.

정리하신다고 고생많으셨습니다. 감사합니다.

**Reply by coalong** [reply]
> 즉 자바의 인터페이스는 추상화를 하는 이유를 결과로 내놓고 있고 ADT에 가장 잘 대응되는 개념인 이유도 알게 되었습니다.
   
좋은 글을 자세히 써주셔서 감사합니다.! 🐱 

책에서는 다중상속을 구현할 때 문제점과 그에 대한 대안 방법만 나와있다면, 더 본질적이고 중요한
추상클래스보다 인터페이스를 우선시 하는 이유를 객체지향 프로그래밍의 추상화 개념에 대하여 써주셔서 많은 도움됐습니다~! 👍👍

**Comment by corock** [comment]
잘 읽었습니다 😄
아이템 제목만 보고 처음에는 가벼운 내용으로 되어있을 줄 알았는데, 내용이 그렇게 쉽지만은 않네요.
그리고 무엇보다 굉장히 시간을 많이 투자한 느낌이 정리하신 내용에서 보입니다 👍

해당 아이템 장을 읽으면서 생각해볼 만한 점을 나열해 보면 다음과 같습니다.

1. Java 8 이후부터 사용할 수 있는 디폴트 메서드(*default method*) 파급력
2. 자바에서 다중 상속을 지원하지 않음으로써 더욱 부각되는 인터페이스
3. 인터페이스가 가지는 특징으로, 레거시 코드에도 쉽게 추가가 가능한 확장성과 계층구조가 없는 타입 프레임워크 생성 가능 등의 유연함
4. 조합 폭발(*combinational explosion*), 추상 골격 구현(*skeletal implementation*) 등 한글 번역이 가져다주는 어색함

**Reply by coalong** [reply]
> 해당 아이템 장을 읽으면서 생각해볼 만한 점을 나열해 보면 다음과 같습니다.

이번 아이템의 핵심 주제는 간단명료한데,  안에 내용들을 이해하고 깔끔하게 정리하기가 어려워 고민했는데,
@corock 님이 적어주신 대로 마지막에 생각해볼 만한 점을 정리해보는 것도 도움이 될 거 같아요! 
감사합니다. 👍
